test/k8s-integration/main.go

```diff
@@ -317,30 +344,37 @@ func handle() error {
 	return nil
 }
 
-func generateTestSkip(normalizedVersion string) string {
+func generateGCETestSkip(clusterVersion string) string {
 	skipString := "\\[Disruptive\\]|\\[Serial\\]"
-	switch normalizedVersion {
-	// Fall-through versioning since all test cases we want to skip in 1.15
-	// should also be skipped in 1.14
-	case "1.13":
-		fallthrough
-	case "1.14":
-		fallthrough
-	case "1.15":
-		fallthrough
-	case "1.16":
-		// "volumeMode should not mount / map unused volumes in a pod" tests a
-		// bug-fix introduced in 1.17
-		// (https://github.com/kubernetes/kubernetes/pull/81163)
+	v := apimachineryversion.MustParseSemantic(clusterVersion)
+
+	// "volumeMode should not mount / map unused volumes in a pod" tests a
+	// (https://github.com/kubernetes/kubernetes/pull/81163)
+	if v.LessThan(apimachineryversion.MustParseSemantic("1.16.0")) {
 		skipString = skipString + "|volumeMode\\sshould\\snot\\smount\\s/\\smap\\sunused\\svolumes\\sin\\sa\\spod"
-		// Skip Snapshot tests pre 1.17
-		skipString = skipString + "|snapshot"
-		fallthrough
-	case "1.17":
-	case "latest":
-	case "master":
-	default:
 	}
+
+	if v.LessThan(apimachineryversion.MustParseSemantic("1.17.0")) {
+		skipString = skipString + "|VolumeSnapshotDataSource"
+	}
+	return skipString
+}
+
+func generateGKETestSkip(clusterVersion string, use_gke_managed_driver bool) string {
+	skipString := "\\[Disruptive\\]|\\[Serial\\]"
+	curVer := mustParseVersion(clusterVersion)
+	if curVer.lessThan(mustParseVersion("1.16.0")) {
+		skipString = skipString + "|volumeMode\\sshould\\snot\\smount\\s/\\smap\\sunused\\svolumes\\sin\\sa\\spod"
+	}
+
+	// For GKE deployed PD CSI snapshot is enabled in 1.17.5-gke.9(and higher), 1.17.6-gke.4(and higher), 1.18.3(and higher).
+	if (use_gke_managed_driver && curVer.lessThan(mustParseVersion("1.17.5-gke.9"))) ||
```

Comment:
Would this read better writen like

if (use_gke_managed_driver) {
  if curVer.lessThan(1.17.5-gke.9) {
    skipString = skipString + "|VolumeSnapshotDataSource"
  } else if curVer.LessThen(1.17.6.0) {
   // don't skip
  } else if curVer.LessThan(17.6-gke.4) {
   skipString = ...
  } else if curVer.LessThan(1.18.0) {
   // don't skip
  } else if curVer.LessThan(1.18.3) {
   skipStyring = ...
 }

It's still complicated but at least one can just read down and quickly see that status of each version rather than trying to figure out more complicated nested logic.

Reply:
I think we also don't need to be so exact with the versions considering that we control the test job configurations too.

Reply:
@msau42  keeping the logic accurate would make sure the test skip strings logic robust and gives clarity which version exactly supports snapshots. Thoughts?

Reply:
I think we can handle the supported versions in the top level README (I doubt anyone will actually come down to the test code to figure it out), and maybe clarify as a comment in the test code if we want to.

On the GKE side, I think we can handle the complexity of only adding the test job for supported versions.

Reply:
From the prow configs I see we add cluster versions like -extract=gke-latest-1.17 (instead of adding specific gke releases)
so for e.g if a version < 1.17.6-gke.4 is picked test breaks. Feel it's easier to control and be deterministic in the go code. Let's discuss on this offline and reach a conclusion.

Reply:
based on @msau42 's recommendation, given the prow jobs ensure to test against latest of any cluster version, keep the logic simple to check for < 1.17.5-gke.9 for skip strings, and added a comment.